### ISSUE TITLE
Image similarity?

### ISSUE THREAD

**Issue by youssefavx** [openai/CLIP#1]
Incredible work as always you guys! In looking at the Colab, it seems it's possible to do image-to-text similarity but I'm curious if it's possible to compare image similarity as well.

For instance, if I just replace 'text_features' with 'image_features' would that work / be the best way to do this?

```
image_features /= image_features.norm(dim=-1, keepdim=True)
image_2_features /= image_2_features.norm(dim=-1, keepdim=True)
similarity = image_2_features.cpu().numpy() @ image_features.cpu().numpy().T
```

**Comment by jongwook** [comment]
Thanks!

I believe image similarities will work, although we haven't experimented/evaluated tasks using image similarity measures specifically.

**Comment by youssefavx** [comment]
Thank you! I'll try it out and see

**Comment by rom1504** [comment]
I tried it and it seems to work pretty well. (actually better than using a pretrained on classification efficientnet for example)

**Comment by woctezuma** [comment]
> I tried it and it seems to work pretty well. (actually better than using a pretrained on classification efficientnet for example)

Just to be clear: when you see it works well (and even better than another method), you mean that you have this feeling based on visual inspection of a few examples, or is it based on an image dataset which you had built beforehand to assess the methods? I would not have any issue with any of the methods, but it is just out of curiosity.

**Comment by rom1504** [comment]
Visual inspection for now (putting image embeddings in a faiss index and trying a few queries), but I also intend to do a more extensive evaluation later on.

I know there are also image retrieval research dataset and tasks where evaluating this model would give the most reliable and comparable results.

**Comment by woctezuma** [comment]
Results are interesting, but one has to be careful because the model likes to pick text from images.

For instance, here, I suspect that images are similar because the word "story" appears.

![Images with the word story](https://i.imgur.com/ITdAbIQ.jpg)

**Comment by scorpionsky** [comment]
Maybe that is because the words (e.g., "story") appear in the text..

**Comment by woctezuma** [comment]
> Maybe that is because the words (e.g., "story") appear in the text..

I use the pre-trained model, and I only show the images it. Zero text.

**Comment by scorpionsky** [comment]
Thanks for the clarification. Then the model might have learned attentions to the embedded text regions during the training. It is very interesting, in both of the cases.

**Comment by woctezuma** [comment]
I have added a search engine to look for specific Steam games:
https://woctezuma.github.io/steam-svelte-autocomplete/index.html

I am still having fun with this.

**Comment by thoppe** [comment]
I made an [app](https://share.streamlit.io/thoppe/streamlit-clip-unsplash-explorer) to explore how image similarity works in CLIP by tying it to the closest images in Unsplash. Like @woctezuma noted, it seems to latch onto images with matching text, but in generally there is a lot more going on:

![Screenshot from 2021-01-27 14-29-05](https://user-images.githubusercontent.com/2707106/106049534-36850880-60b4-11eb-97a0-c1196f7820f8.png)

![Screenshot from 2021-01-27 14-35-48](https://user-images.githubusercontent.com/2707106/106049552-3a188f80-60b4-11eb-9daf-eed707283185.png)

![Screenshot from 2021-01-27 14-38-37](https://user-images.githubusercontent.com/2707106/106049556-3b49bc80-60b4-11eb-8df4-1fe83a01b8f8.png)

**Comment by htoyryla** [comment]
Anyone has any experience on using CLIP image encoding in a loss function (instead pixel loss like MSE, perceptional loss like LPIPS or structural similar like SSIM) to guide image generation (for instance, using generating images matching both text and a given image) or, for example, VAE training. My experiments with this have not been successful, it simply does not converge. I wonder if it is the case that CLIP image encoding is strictly semantic and does not contain enough visual features to properly guide training to produce images which resemble, for instance, compositionally to a given image.

**Comment by thoppe** [comment]
@htoyryla there is an active community on twitter trying to do just that! I found an interesting point about the text vectors vs image vectors -- they aren't colocated!

https://twitter.com/metasemantic/status/1356406256802607112

**Comment by htoyryla** [comment]
Thanks...  my problem seems BTW more like finding the image from the image encoding.

**Comment by woctezuma** [comment]
> Anyone has any experience on using CLIP image encoding in a loss function (instead pixel loss like MSE, perceptional loss like LPIPS or structural similar like SSIM) to guide image generation (for instance, using generating images matching both text and a given image)

https://github.com/orpatashnik/StyleCLIP

**Comment by htoyryla** [comment]
Oh, sorry, it looks like I formulated my question ambiguously. I am specifically looking for a case which uses CLIP to compare similarity _between two images_, i.e. loss calculated from two image embeddings instead of using a more conventional image loss (MSE, LPIPS or SSIM), possibly together with CLIP to compare text with image.

The above arrangement uses CLIP in the normal way to compare text and image embeddings.

**Comment by woctezuma** [comment]
Relevant to some discussion earlier in the thread (and at least an interesting read): https://distill.pub/2021/multimodal-neurons/

**Comment by Sxela** [comment]
> Results are interesting, but one has to be careful because the model likes to pick text from images.
> 
> For instance, here, I suspect that images are similar because the word "story" appears. The query is the top left picture. In the second to last picture, there is the word "stonies". It is only in the last picture that there is a visually similar planet shape in the same position as in the query.
> 
> ![Images with the word story](https://camo.githubusercontent.com/d95ddbc85cb5982b1aa5fc7ea264dc5c00e5044a74d558669f6d16285f38dae2/68747470733a2f2f692e696d6775722e636f6d2f495464416249512e6a7067)
> 
> Similarly, here, it is the word "empire" which seems to drive the matching between these images.
> 
> ![Images with the word empire](https://camo.githubusercontent.com/c7fd3044dac1726884c7587109f272ac13f0e57fac464f4050a309edc8ee824e/68747470733a2f2f692e696d6775722e636f6d2f6a586472366e722e6a7067)
> 
> Here, the word "forest". Etc.
> 
> ![Images with the word forest](https://camo.githubusercontent.com/7a9a76d28cde5ddc0a68eddd4e1e1143e3c8e35d31252f7acd1f4aac8b5991bd/68747470733a2f2f692e696d6775722e636f6d2f507341676630362e6a7067)
> 
> It is both impressive... and a bit disappointing if I want to use the model for retrieving similar **images**.

You can actually remove text (via some text detector) from those images before feeding them to clip and see how it goes.

**Comment by matrixgame2018** [comment]
Hello， I just want to know  the L2-norm and cos will be working in the case which between the image and image ？

**Comment by woctezuma** [comment]
Slightly relevant: this figure in [the `DALL-E 2` paper](https://cdn.openai.com/papers/dall-e-2.pdf).

![Picture](https://i.imgur.com/SsuBjgu.png

**Comment by FurkanGozukara** [comment]
can anyone post a fully working code sample (a google colab notebook would be amazing) that will take image A and image B and calculate their image similarity?

Thank you so much

@woctezuma @iremonur @matrixgame2018 @Sxela @htoyryla @thoppe @scorpionsky @rom1504 @youssefavx @jongwook

**Comment by woctezuma** [comment]
First, you extract features as shown [in the README](https://github.com/openai/CLIP#usage):

```python
import torch
import clip
from PIL import Image

device = "cuda" if torch.cuda.is_available() else "cpu"
model, preprocess = clip.load("ViT-B/32", device=device)

image = preprocess(Image.open("CLIP.png")).unsqueeze(0).to(device)

with torch.no_grad():
    image_features = model.encode_image(image)
```

Then you compute similarity between features, e.g. as [in the original post](https://github.com/openai/CLIP/issues/1#issue-779675076).

**Comment by FurkanGozukara** [comment]
> First, you extract features as shown [in the README](https://github.com/openai/CLIP#usage):
> 
> ```python
> import torch
> import clip
> from PIL import Image
> 
> device = "cuda" if torch.cuda.is_available() else "cpu"
> model, preprocess = clip.load("ViT-B/32", device=device)
> 
> image = preprocess(Image.open("CLIP.png")).unsqueeze(0).to(device)
> 
> with torch.no_grad():
>     image_features = model.encode_image(image)
> ```
> 
> Then you compute similarity between features, e.g. as [in the original post](https://github.com/openai/CLIP/issues/1#issue-779675076) or as [in the README](https://github.com/openai/CLIP#more-examples):
> 
> ```python
> # Pick the top 5 most similar labels for the image
> image_features /= image_features.norm(dim=-1, keepdim=True)
> text_features /= text_features.norm(dim=-1, keepdim=True)
> similarity = (100.0 * image_features @ text_features.T).softmax(dim=-1)
> ```

thanks a lot for answer

i did come up with this way what you think?

also do you think clip is the best available method atm for such task? 
lets say i liked a jacket on a person and i want to find similar jackets from my database 

```
import torch
from transformers import CLIPImageProcessor, CLIPModel, CLIPTokenizer
from PIL import Image

# Load the CLIP model
model_ID = "openai/clip-vit-large-patch14"
model = CLIPModel.from_pretrained(model_ID)

preprocess = CLIPImageProcessor.from_pretrained(model_ID)

# Define a function to load an image and preprocess it for CLIP
def load_and_preprocess_image(image_path):
    # Load the image from the specified path
    image = Image.open(image_path)

    # Apply the CLIP preprocessing to the image
    image = preprocess(image, return_tensors="pt")

    # Return the preprocessed image
    return image

# Load the two images and preprocess them for CLIP
image_a = load_and_preprocess_image('/content/a.JPG')["pixel_values"]
image_b = load_and_preprocess_image('/content/e.png')["pixel_values"]

# Calculate the embeddings for the images using the CLIP model
with torch.no_grad():
    embedding_a = model.get_image_features(image_a)
    embedding_b = model.get_image_features(image_b)

# Calculate the cosine similarity between the embeddings
similarity_score = torch.nn.functional.cosine_similarity(embedding_a, embedding_b)

# Print the similarity score
print('Similarity score:', similarity_score.item())
```

**Comment by woctezuma** [comment]
Sorry, I don't have any authority on this subject. I cannot help more or recommend a method over another. 😅 

That being said, out of curiosity, I would probably try [`X-Decoder`](https://x-decoder-vl.github.io/).

> Zou, Xueyan, et al. "Generalized Decoding for Pixel, Image, and Language." arXiv preprint [arXiv:2212.11270](https://arxiv.org/abs/2212.11270) (2022).
